tests/base/test_alarm.py

```diff
@@ -357,12 +386,100 @@ def test_reset(self):
+                self.assertEqual(alarm.timestamp_unmute, 0)
+
+    async def test_mute_invalid(self):
+        good_user = "otho"
+        failed_user = "user associated with invalid mute command"
+        good_duration = 5
+        good_severity = AlarmSeverity.WARNING
+        for alarm in self.alarm_iter(name=good_user, callback=None):
+            for bad_duration, bad_severity in itertools.product((0, -0.01), (AlarmSeverity.NONE, -53)):
+                # check that mute raises ValueError for invalid values
+                # and leaves the alarm state unchanged
+                initial_alarm = copy.copy(alarm)
+                with self.assertRaises(ValueError):
+                    alarm.mute(duration=bad_duration, severity=good_severity, user=failed_user)
+                with self.assertRaises(ValueError):
+                    alarm.mute(duration=good_duration, severity=bad_severity, user=failed_user)
+                with self.assertRaises(ValueError):
+                    alarm.mute(duration=bad_duration, severity=bad_severity, user=failed_user)
+                self.assertEqual(alarm, initial_alarm)
+
+                # make sure failures also leave muted alarm state unchanged
+                alarm.mute(duration=good_duration, severity=good_severity, user=good_user)
+                muted_alarm = copy.copy(alarm)
+                self.assertTrue(alarm.muted)
+                self.assertEqual(alarm.muted_by, good_user)
+                self.assertEqual(alarm.muted_severity, good_severity)
+
+                with self.assertRaises(ValueError):
+                    alarm.mute(duration=bad_duration, severity=good_severity, user=failed_user)
+                with self.assertRaises(ValueError):
+                    alarm.mute(duration=good_duration, severity=bad_severity, user=failed_user)
+                with self.assertRaises(ValueError):
+                    alarm.mute(duration=bad_duration, severity=bad_severity, user=failed_user)
+                self.assertEqual(alarm, muted_alarm)
+
+                alarm.unmute()  # kill unmute timer
+
+    async def test_unmute(self):
+        user = "otho"
+        duration = 5
+        for severity in AlarmSeverity:
+            if severity == AlarmSeverity.NONE:
+                continue  # invalid value
+            for alarm in self.alarm_iter(name=user, callback=self.callback):
+                self.ncalls = 0
+                # check that unmute on unmuted alarm is a no-op
+                original_alarm = copy.copy(alarm)
+                alarm.unmute()
+                self.assertEqual(alarm, original_alarm)
+                self.assertEqual(self.ncalls, 1)
+
+                # mute alarm and unmute it again before it unmutes itself
+                alarm.mute(duration=duration, severity=severity, user=user)
+                curr_tai = salobj.tai_from_utc(time.time())
+                self.assertEqual(self.ncalls, 2)
```

Comment:
Ah, now I understand what that thing in `setUp` is doing. That's rather a rather clever use of testing the callback.